rotkehlchen/tests/exchanges/test_binance.py

```diff
@@ -790,3 +768,67 @@ def mock_get_deposit_withdrawal(url):  # pylint: disable=unused-argument
                 end_ts=Timestamp(end_ts),
             )
             assert mock_api_query_dict.call_args_list == expected_calls
+
+
+@pytest.mark.freeze_time(datetime(2020, 11, 24, 3, 14, 15))
+def test_api_query_retry_on_status_code_429(function_scope_binance):
+    """Test when Binance API returns 429 and the request is retried, the
+    signature is not polluted by any attribute from the previous call.
+
+    It also tests getting the `retry-after` seconds to backoff from the
+    response header.
+
+    NB: basically remove `call_options['signature']`.
+    """
+    binance = function_scope_binance
+    offset_ms = 1000
+    call_options = {
+        'fromId': 0,
+        'limit': 1000,
+        'symbol': 'BUSDUSDT',
+        'recvWindow': 10000,
+        'timestamp': str(ts_now_in_ms() + offset_ms),
+    }
+    signature = hmac.new(
+        binance.secret,
+        urlencode(call_options).encode('utf-8'),
+        hashlib.sha256,
+    ).hexdigest()
+    call_options['signature'] = signature
+    base_url = 'https://api.binance.com/api/v3/myTrades?'
+    exp_request_url = base_url + urlencode(call_options)
+
+    # NB: both calls must have the same signature (time frozen)
+    expected_calls = [call(exp_request_url), call(exp_request_url)]
+
+    def get_mocked_response():
+        responses = [
+            MockResponse(429, '[]', headers={'retry-after': '1'}),
+            MockResponse(418, '{"code": "XXX", "msg": "msg"}'),
+        ]
+        for response in responses:
+            yield response
+
+    def mock_response(url):  # pylint: disable=unused-argument
+        return next(get_response)
+
+    get_response = get_mocked_response()
+
+    # NB: force to break the loop after 2 requests by lowering the `self.backoff_limit`
+    backoff_limit_ = MagicMock(return_value=7)
+    offset_ms_ = MagicMock(return_value=1000)
+    session_get = MagicMock(side_effect=mock_response)
+    with patch.object(binance, 'backoff_limit', new_callable=backoff_limit_):
```

Comment:
hahha that's quite a deeply nested context manager inception.

You should be using the ExitStack in such cases: https://github.com/rotki/rotki/blob/f3a795e0a3061751db22e6ebfcf7d4701488878d/rotkehlchen/tests/api/test_balances.py#L161-L162

Reply:
Done.